GenericList/string_list.cpp

```diff
@@ -56,50 +104,93 @@ std::string List_String::ToString() {
 	return result;
 }
 
-
-bool List_String::Add(std::string new_val) {
-	if (capacity < count + 1) {
-		Capacity(capacity == 0 ? 1 : capacity * 2); //double array size
+std::string* List_String::Find(const std::string& val) {
+	for (int i = 0; i < count; i++) {
+		if (data[i] == val) return data+i; //TODO: research why &data[i] might get overloaded
 	}
-	data[count++] = new_val;
-	return true;
+	return nullptr;
 }
 
-bool List_String::Contains(std::string val) {
-	return IndexOf(val) != -1;
+
+const std::string* List_String::Find(const std::string& val) const {
+	for (int i = 0; i < count; i++) {
+		if (data[i] == val) return data + i; //TODO: research why &data[i] might get overloaded
+	}
+	return nullptr;
 }
 
-bool List_String::RemoveAt(int index) {
-	if (index < 0) return false;
-	if (index >= count) return false;
-	//allowed setting: index = [0, count-1], count > 0 (data is initialized)
 
-	std::memmove(data + index, data + index + 1, count - index - 1);
-	count--;
-	return true;
+void List_String::RemoveAt(int index) {
+	if (index < 0 || index >= count) throw std::out_of_range(string("Cannot remove element at out_of_range index: ") + to_string(index) + string("."));
+	
+	std::move(data + index + 1, data + count, data + index);
+	data[--count].~string();
 }
 
 
-int List_String::IndexOf(std::string val) {
-	for (int i = 0; i < count; i++) {
-		if (data[i] == val) return i;
-	}
-	return -1;
+
+bool List_String::Remove(const std::string& val) {
+	std::string* ptr = Find(val);
+	if (ptr!=nullptr) RemoveAt(ptr-data);
+	return ptr!=nullptr;
 }
 
 
-bool List_String::Remove(std::string val) {
-	int index = IndexOf(val);
-	return RemoveAt(index);
+
+const std::string& List_String::Get(int index) const {
+	if (index >= count || index < 0) throw std::out_of_range(string("Cannot get element at out_of_range index: ") + to_string(index) + string(")"));
+
+	return data[index];
 }
 
-std::string List_String::Get(int index) {
+std::string& List_String::Get(int index) {
 	if (index >= count || index < 0) throw std::out_of_range(string("Cannot get element at out_of_range index: ") + to_string(index) + string(")"));
 
 	return data[index];
 }
 
 
+std::string* List_String::CreateDeepCopy(std::string* data, size_t data_size, size_t copy_size) {
+	if (copy_size > data_size || copy_size < 0) {
+		throw std::out_of_range(string("Cannot copy array of size ") + to_string(copy_size) + string(" onto array of size ") + to_string(copy_size) + string("."));
+	}
```

Comment:
Just noticed that this function is private, so you can just use `assert` here as well instead of throwing the exception. If the input is incorrect, it's an issue on your end as the implementer, not the user, so you shouldn't throw an exception about an error the user can do nothing about.